### ISSUE TITLE
Command to upgrade a cluster

### ISSUE THREAD

**Issue by marians** [giantswarm/gsctl#187]
Part of https://github.com/giantswarm/giantswarm/issues/2022

To manually trigger an upgrade of a guest cluster to a newer version, gsctl needs a command.
 
The backend automatically selects to which version the cluster will be upgraded. The logic for that can be replicated like this:

- Pick the lowest version number that is higher than the current version number.
- Pick only from active releases.

This can be used to display to the user to which version we'll upgrade and what the changes are.

In addition, a confirmation should be required by default, overridable using `-f/--force`.

**Comment by marians** [comment]
The most straightforward way to name the command:

    gsctl upgrade cluster

Details

- Details of the target version and a basic comparison between the current and the target version as well as the target version's changelog will be displayed. See mockup below.
- A confirmation will be required unless the `-f` or `--force` flag has been added.

(WIP, will be edited)

**Comment by oponder** [comment]
Looks good to me! I am missing the "warning" about what the consequences of the upgrade are (workers will be drained, etc). Are we consciously excluding that step from `gsctl`?

**Comment by marians** [comment]
Yes, good that you ask. With the general assumption of gsctl being the powertool for knowledgeable users (who feel that happa might be too chatty, to verbose, too slow), I decided to not show that. It would make the confirmation step content really big. And I really don't want a two step confirmation in gsctl.

**Comment by oponder** [comment]
IMHO, powertools should still have safety features or some warnings on them.

I think the user must at least be aware of these two things:

1. Make sure your services still run with one node fully drained.
2. The kubernetes API will be unavailable for some time during the upgrade.

Perhaps we can add just these two notices?

Alternatively, I came up with an idea: what if we temporarily add a worker as part of the upgrade process.

That way we create the necessary space to successfully perform an upgrade. I know this is very not in scope of this discussion though.

**Comment by oponder** [comment]
Reminded me of this woodworking video I saw. The kid asks "Why does the start button have a cover over it, but the stop button doesn't"?

Then the dad says "So you can't accidentally hit the start button, but you want to be able to hit the stop button very quickly"

<img width="903" alt="screen shot 2018-01-10 at 17 20 20" src="https://user-images.githubusercontent.com/455309/34783086-daa0485a-f62a-11e7-82d0-0a16b8d94e4c.png">

<img width="888" alt="screen shot 2018-01-10 at 17 20 57" src="https://user-images.githubusercontent.com/455309/34783121-f5e23f42-f62a-11e7-943f-d99ce7845777.png">

**Comment by marians** [comment]
Alternative that adds a warning text on upgrade side effects, linking to a detailed document:

```
$ gsctl upgrade cluster clusterid
Cluster 'clusterid' will be upgraded from version 1.4.7 to 1.5.0.

Component changes:

    kubernetes: Upgrade from 1.8.1 to 1.8.3
    docker: Added (version 1.12.6)
    calico: Removed

Changelog:

    - Kubernetes upgraded to version 1.8.3, to fix performance issues and a security problem.
    - docker added
    - calico removed

NOTE: Upgrading may impact your running workloads and will make the cluster's
Kubernetes API unavailable temporarily. Before upgrading, please acknowledge the
details described in

    https://docs.giantswarm.io/reference/cluster-upgrades/

Do you want to start the upgrade now? [y/n] 
```

**Comment by marians** [comment]
Idea is to change the logic to have the command available to admins, then merge.

**Comment by marians** [comment]
@oponder Any idea what would be the most elegant way to determine whether a user is admin? What's the logic we use in happa? )might make sense to use the same here)

**Comment by oponder** [comment]
Right now, in happa, it's based on wether or not the user can see the `giantswarm` organisation.  (so basically if `giantswarm` is present in /v4/organizations/`

I say `can see` because the user does not have to actually be a member of the `giantswarm` organisation. Theres some interesting theoretical edge cases possible due to this, and we're kinda lucky that they don't effect us usually.

One consequence for example is that it's impossible for happa to know if someone is an admin or not when there is no `giantswarm` organization in the system. But this would never happen normally.

As we start moving further, and dropping the `giantswarm` token, we would determine a users rights based on the `scopes` of their access token.

**Comment by marians** [comment]
The upgrade command has been shipped with 0.14.0.
